lib/floe/workflow.rb

```diff
@@ -41,18 +41,19 @@ def step
 
       input = @context["State"]["Output"] || @context["Execution"]["Input"].dup
 
+      @context["State"] = {
+        "EnteredTime" => Time.now,
+        "Input"       => input,
+        "Name"        => current_state.name
+      }
+
       tick = Process.clock_gettime(Process::CLOCK_MONOTONIC)
       next_state, output = current_state.run!(input)
       tock = Process.clock_gettime(Process::CLOCK_MONOTONIC)
 
-      @context["State"] = {
-        "EnteredTime"  => tick,
-        "FinishedTime" => tock,
-        "Duration"     => tock - tick,
-        "Output"       => output,
-        "Name"         => next_state&.name,
-        "Input"        => output
-      }
+      @context["State"]["FinishedTime"] = Time.now
+      @context["State"]["Duration"]     = (tock - tick) / 1_000_000.0
+      @context["State"]["Output"]       = output
 
       @context["States"] << @context["State"]
```

Comment:
should this be the `State` that we just built, or the previous state?

Reply:
do we need to save these when there is an error?

Reply:
> should this be the State that we just built, or the previous state?

This is the state that we're currently running

> do we need to save these when there is an error?

That is a good point I do need more robust error handling here.  The state run methods do handle exceptions already so they can execute the Fail conditions but if something is non-recoverable it might raise up here.